bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyScrollingControl.java

```diff
@@ -0,0 +1,482 @@
+import org.eclipse.swt.widgets.Label;
+
+import org.eclipse.core.runtime.ICoreRunnable;
+import org.eclipse.core.runtime.jobs.Job;
+
+import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.layout.GridLayoutFactory;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.ITextPresentationListener;
+import org.eclipse.jface.text.ITextViewerExtension;
+import org.eclipse.jface.text.ITextViewerExtension4;
+import org.eclipse.jface.text.ITextViewerExtension5;
+import org.eclipse.jface.text.JFaceTextUtil;
+import org.eclipse.jface.text.source.CompositeRuler;
+import org.eclipse.jface.text.source.ISourceViewer;
+import org.eclipse.jface.text.source.IVerticalRuler;
+import org.eclipse.jface.text.source.IVerticalRulerColumn;
+
+import org.eclipse.ui.internal.texteditor.LineNumberColumn;
+
+/**
+ * This class builds a control that is rendered on top of the given source viewer. The controls
+ * shows the sticky lines that are set via {@link #setStickyLines(List)} on top of the source
+ * viewer. The {@link StickyLine#lineNumber()} is linked to to corresponding line number in the
+ * given source viewer, with index starting at 0.
+ * 
+ * As part of its responsibilities, the class handles layout arrangement and styling of the sticky
+ * lines along with navigation to the respective sticky line.
+ *
+ * Note: The dispose method should be called to clean up system resources associated with this
+ * object when it's no longer needed.
+ */
+public class StickyScrollingControl {
+
+	private List<StickyLine> stickyLines;
+
+	private ISourceViewer sourceViewer;
+
+	private IVerticalRuler verticalRuler;
+
+	private StickyScrollingControlSettings settings;
+
+	private Canvas stickyLinesCanvas;
+
+	private StyledText stickyLineNumber;
+
+	private StyledText stickyLineText;
+
+	private ITextPresentationListener textPresentationListener;
+
+	private ControlListener controlListener;
+
+	private StickyScollingCaretListener caretListener;
+
+	private Label bottomSeparator;
+
+	public StickyScrollingControl(ISourceViewer sourceViewer, StickyScrollingControlSettings settings) {
+		this(sourceViewer, null, settings);
+	}
+
+	public StickyScrollingControl(ISourceViewer sourceViewer, IVerticalRuler verticalRuler, StickyScrollingControlSettings settings) {
+		this.stickyLines= new ArrayList<>();
+		this.sourceViewer= sourceViewer;
+		this.verticalRuler= verticalRuler;
+		this.settings= settings;
+
+		createControls();
+		addSourceViewerListeners();
+	}
+
+	/**
+	 * Sets the sticky lines to show. The line numbers are linked to the line number in the
+	 * corresponding source viewer, starting with index 0.
+	 * 
+	 * @param stickyLines The sticky lines to show
+	 */
+	public void setStickyLines(List<StickyLine> stickyLines) {
+		if (!stickyLines.equals(this.stickyLines)) {
+			this.stickyLines= stickyLines;
+			updateStickyScrollingControls();
+		}
+	}
+
+	public void setSettings(StickyScrollingControlSettings settings) {
```

Comment:
```suggestion
	public void applySettings(StickyScrollingControlSettings settings) {
```

But that's just preference ;)